docs/development/CUSTOM_RULES.md

```diff
@@ -0,0 +1,206 @@
+
+That `parserServices` object contains:
+
+- `program`: A full TypeScript `ts.Program` object
+- `esTreeNodeToTSNodeMap`: Map of `@typescript-eslint/estree` `TSESTree.Node` nodes to their TypeScript `ts.Node` equivalents
+- `tsNodeToESTreeNodeMap`: Map of TypeScript `ts.Node` nodes to their `@typescript-eslint/estree` `TSESTree.Node` equivalents
+
+By mapping from ESTree nodes to TypeScript nodes and retrieving the TypeScript program from the parser services, rules are able to ask TypeScript for full type information on those nodes.
+
+This rule bans for-of looping over an enum by using the type-checker via typescript-eslint and TypeScript APIs:
+
+```ts
+import { ESLintUtils } from '@typescript-eslint/experimental-utils';
+import * as ts from 'typescript';
+import * as tsutils from 'tsutils';
+
+export const rule: eslint.Rule.RuleModule = {
+  create(context) {
+    return {
+      ForOfStatement(node) {
+        // 1. Grab the TypeScript program from parser services
+        const parserServices = ESLintUtils.getParserServices(context);
+        const checker = parserServices.program.getTypeChecker();
+
+        // 2. Find the backing TS node for the ES node, then that TS type
+        const originalNode = parserServices.esTreeNodeToTSNodeMap.get(
+          node.right,
+        );
+        const nodeType = checker.getTypeAtLocation(node);
+
+        // 3. Check the TS node type using the TypeScript APIs
+        if (tsutils.isTypeFlagSet(nodeType, ts.TypeFlags.EnumLike)) {
+          context.report({
+            messageId: 'loopOverEnum',
+            node: node.right,
+          });
+        }
+      },
+    };
+  },
+  meta: {
+    docs: {
+      category: 'Best Practices',
+      description: 'Avoid looping over enums.',
+    },
+    messages: {
+      loopOverEnum: 'Do not loop over enums.',
+    },
+    type: 'suggestion',
+    schema: [],
+  },
+};
+```
+
+## Testing
+
+`@typescript-eslint/experimental-utils` exports its own `RuleTester` with a similar API to the built-in [ESLint `RuleTester`](https://eslint.org/docs/developer-guide/nodejs-api#ruletester).
+It should be provided with the same `parser` and `parserOptions` you would use in your ESLint configuration.
+
+For rules that don't need type information, passing just the `parser` will do:
+
+```ts
+import { ESLintUtils } from '@typescript-eslint/experimental-utils';
+import rule from './my-rule';
+
+const ruleTester = new ESLintUtils.RuleTester({
+  parser: '@typescript-eslint/parser',
+});
+
+ruleTester.run('my-rule', rule {
+  valid: [/* ... */],
+  invalid: [/* ... */],
+})
+```
+
+### Testing Typed Rules
+
+For rules that do need type information, `parserOptions` must be passed in as well.
+Tests must have at least an absolute `tsconfigRootDir` path provided as well as a relative `project` path from that directory:
```

Comment:
this probably needs to make clear that (right now) we ***need*** to have some physical files on disk for this!

it essentially expects the following files:
- `tsconfig.json` -> tsconfig used as the test "project"
- `file.ts` -> blank test file used for normal TS tests
- `file.tsx` -> blank test file used for tests with `parserOptions: { ecmaFeatures: { jsx: true }}}`